### ISSUE TITLE
question about release_conn() in the presence of unread response data

### ISSUE THREAD

**Issue by christopherhesse** [urllib3/urllib3#1891]
While investigating another issue (https://github.com/urllib3/urllib3/issues/1878), I noticed some odd exceptions from urllib3, and I was wondering what would cause this behavior.

Here's `http.client`:

```
import http.client
conn = http.client.HTTPSConnection("httpbin.org")
conn.request("GET", "/bytes/1024")
r1 = conn.getresponse()
r1.read(4)
conn.request("GET", "/bytes/1024")
r1 = conn.getresponse()
r1.read(4)
```

This produces `http.client.ResponseNotReady: Request-sent`, which makes sense because the response was not entirely read so the connection cannot be reused.

Here's `urllib3`:

```
import urllib3
import logging
logging.basicConfig()
http = urllib3.PoolManager()
r = http.request(
   'GET',
  'http://httpbin.org/bytes/1024',
   preload_content=False)
r.read(4)
r.release_conn()
r = http.request(
   'GET',
  'http://httpbin.org/bytes/1024',
   preload_content=False)
r.read(4)
r.release_conn()
```

```
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): httpbin.org:80
DEBUG:urllib3.connectionpool:http://httpbin.org:80 "GET /bytes/1024 HTTP/1.1" 200 1024
DEBUG:urllib3.util.retry:Incremented Retry for (url='/bytes/1024'): Retry(total=2, connect=None, read=None, redirect=None, status=None)
WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', ResponseNotReady('Request-sent'))': /bytes/1024
DEBUG:urllib3.connectionpool:Starting new HTTP connection (2): httpbin.org:80
DEBUG:urllib3.connectionpool:http://httpbin.org:80 "GET /bytes/1024 HTTP/1.1" 200 1024
```

This produces no error, but logs a warning: `WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', ResponseNotReady('Request-sent'))': /bytes/1024`

This also makes sense, since the pool will create a new connection when it notices that one is busted.

If I run about 100 copies of this program at the same time in a loop:

```
import urllib3
import logging

logging.basicConfig()


http = urllib3.PoolManager()


def fn():
    while True:
        try:
            r = http.request(
                "GET",
                "http://localhost:8081/meow",
                preload_content=False,
                retries=False,
                redirect=False,
            )
            r.read(4)
            r.release_conn()
        except urllib3.exceptions.NewConnectionError:
            continue
        except urllib3.exceptions.ProtocolError as e:
            print(type(e), e)
            continue
        break

def main():
    for i in range(100):
        print(i)
        fn()

if __name__ == "__main__":
    main()

```

Against this go server:

```
package main

import (
    "fmt"
	"log"
	"strings"
    "net/http"
)

func main() {
	fmt.Println("starting")
	body := strings.Repeat("meow\n", 250000)

    http.HandleFunc("/meow", func(w http.ResponseWriter, r *http.Request){
		fmt.Fprintf(w, body)
    })

	fmt.Println("serving")
    log.Fatal(http.ListenAndServe("localhost:8081", nil))
}
```

I get output like this:

```
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
<class 'urllib3.exceptions.ProtocolError'> ('Connection aborted.', BadStatusLine('\n'))
18
19
20
21
22
23
24
25
26
27
28
<class 'urllib3.exceptions.ProtocolError'> ('Connection aborted.', BadStatusLine('\n'))
29
30
```

It seems like the file contents are being interpreted as a status line, but I don't see how that could happen.  Any ideas?

In addition, against a different server (`fake-gcs-server`) I also see `<class 'urllib3.exceptions.ProtocolError'> ('Connection aborted.', ResponseNotReady('Request-sent'))` which implies that the response not ready error is escaping out of the pool somehow.

**Comment by sigmavirus24** [comment]
> This produces `http.client.ResponseNotReady: Request-sent`, which makes sense because the response was not entirely read so the connection cannot be reused.

Generally speaking, the use case of reading a couple of bytes and then discarding the connection isn't something the library is designed for. For you to manually release the connection, you should know that the body is finished.

**Comment by sethmlarson** [comment]
Maybe we should document somewhere that the HTTP response body should be flushed before manually calling `.release_conn()`? Or could `.release_conn()` be amended to automatically flush the body if that hasn't been done yet? That section of our HTTPResponse API is pretty clunky and gotcha-prone for users.

**Comment by christopher-hesse** [comment]
It seems pretty reasonable to not support this use case, and (for instance) to always throw an error on `release_conn()` when the user does this or to just discard the connection if `release_conn()` receives one where the body has not been read (provided that is possible to detect).

The requirement that the body be finished makes sense but is not mentioned in the `release_conn()` docs: https://urllib3.readthedocs.io/en/latest/advanced-usage.html?highlight=release_conn#streaming-and-io and in fact there's even an example implying that you can release a connection that is not finished:

```
>>> import io
>>> reader = io.BufferedReader(r, 8)
>>> reader.read(4)
>>> r.release_conn()
```

@sethmlarson by `flushed` do you mean that `release_conn()` would read until the response was completely downloaded?

**Comment by sigmavirus24** [comment]
> @sethmlarson by `flushed` do you mean that `release_conn()` would read until the response was completely downloaded?

I believe that's what he's saying, yes.

**Comment by christopher-hesse** [comment]
I no longer use `release_conn()` at all to avoid weird issues, but if anyone else is using `release_conn()` for streaming or very large HTTP responses that are meant to be truncated, wouldn't that cause their programs to hang?

**Comment by sigmavirus24** [comment]
> but if anyone else is using `release_conn()` for streaming or very large HTTP responses that are meant to be truncated, wouldn't that cause their programs to hang?

Yes, but the alternative is for that socket to not close properly and hang open causing other issues.

**Comment by christopher-hesse** [comment]
Seems like that behavior should be opt-in then to avoid breaking existing programs (assuming anyone besides me relied on the previous behavior).  For the default behavior, it would make sense to me to close the socket and issue a warning, which seems to be what is done now most of the time.  Is it possible to detect this via `http.client.ResponseNotReady` or some other means?

**Comment by christopherhesse** [comment]
Assuming this won't be changed in `urllib3`, is there an easy way to ensure a file has been completely read before calling `release_conn()`?  I think there would be a performance improvement for when users of my library are reading the entire file (which certainly happens sometimes).
